File: system-test/src/testkit/dsl/core/usecase/BookShopDsl.ts
import type { BookShopDriver } from '../../../driver/port/book-shop-driver.js';
import type { UseCaseContext } from '../shared/use-case-context.js';
import { PlaceOrder } from './usecases/PlaceOrder.js';
import { ViewOrder } from './usecases/ViewOrder.js';
import { CancelOrder } from './usecases/CancelOrder.js';
import { DeliverOrder } from './usecases/DeliverOrder.js';
import { BrowseCoupons } from './usecases/BrowseCoupons.js';
import { PublishCoupon } from './usecases/PublishCoupon.js';
import { GoToBookShop } from './usecases/GoToBookShop.js';

export class BookShopDsl {
  constructor(
    private readonly driver: BookShopDriver,
    private readonly context: UseCaseContext,
  ) {}

  goToBookShop(): GoToBookShop {
    return new GoToBookShop(this.driver, this.context);
  }

  placeOrder(): PlaceOrder {
    return new PlaceOrder(this.driver, this.context);
  }

  viewOrder(): ViewOrder {
    return new ViewOrder(this.driver, this.context);
  }

  cancelOrder(): CancelOrder {
    return new CancelOrder(this.driver, this.context);
  }

  deliverOrder(): DeliverOrder {
    return new DeliverOrder(this.driver, this.context);
  }

  browseCoupons(): BrowseCoupons {
    return new BrowseCoupons(this.driver, this.context);
  }

  publishCoupon(): PublishCoupon {
    return new PublishCoupon(this.driver, this.context);
  }

  async close(): Promise<void> {
    await this.driver.close();
  }
}

File: system-test/src/main/java/com/mycompany/bookshop/testkit/dsl/core/usecase/BookShopDsl.java
package com.jcupac.bookshop.testkit.dsl.core.usecase;

import com.jcupac.bookshop.testkit.driver.port.BookShopDriver;
import com.jcupac.bookshop.testkit.dsl.core.usecase.usecases.BrowseCoupons;
import com.jcupac.bookshop.testkit.dsl.core.usecase.usecases.CancelOrder;
import com.jcupac.bookshop.testkit.dsl.core.usecase.usecases.DeliverOrder;
import com.jcupac.bookshop.testkit.dsl.core.usecase.usecases.GoToBookShop;
import com.jcupac.bookshop.testkit.dsl.core.usecase.usecases.PlaceOrder;
import com.jcupac.bookshop.testkit.dsl.core.usecase.usecases.PublishCoupon;
import com.jcupac.bookshop.testkit.dsl.core.usecase.usecases.ViewOrder;
import com.jcupac.bookshop.testkit.common.Closer;
import com.jcupac.bookshop.testkit.dsl.core.shared.UseCaseContext;

import java.io.Closeable;

public class BookShopDsl implements Closeable {
    private final BookShopDriver driver;
    private final UseCaseContext context;

    public BookShopDsl(BookShopDriver driver, UseCaseContext context) {
        this.driver = driver;
        this.context = context;
    }

    @Override
    public void close() {
        Closer.close(driver);
    }

    public GoToBookShop goToBookShop() {
        return new GoToBookShop(driver, context);
    }

    public PlaceOrder placeOrder() {
        return new PlaceOrder(driver, context);
    }

    public CancelOrder cancelOrder() {
        return new CancelOrder(driver, context);
    }

    public DeliverOrder deliverOrder() {
        return new DeliverOrder(driver, context);
    }

    public ViewOrder viewOrder() {
        return new ViewOrder(driver, context);
    }

    public PublishCoupon publishCoupon() {
        return new PublishCoupon(driver, context);
    }

    public BrowseCoupons browseCoupons() {
        return new BrowseCoupons(driver, context);
    }
}
